### DISCUSSION TITLE
Please assist- Axum, tower-http, ServeDir:  External javascript file referenced from index.html '404 Not Found'

### DISCUSSION THREAD

**Discussion by xabra** [tokio-rs/axum#2874] [Q&A] [answered]
### Summary

I have a very simple Axum server 
It serves the _assets_ directory containing _index.html_
_index.html_  references an external javascript file _my-script.js_,  also in the _assets_ folder

I get 404 not found error for _my-script.js_:
_Loading failed for the <script> with source “http://127.0.0.1:3000/my-script.js”_
It clearly can't find the external javascript file. (Although it works fine if the javascript is inline inside index.html.)

Problem doesn't occur when the page is served using VSCode live server.

What am i doing wrong?

Thanks!

=== src/main.rs ===
``` rust
async fn main() -> Result<()> {
    // Create top level router
    let routes_all = Router::new()
        .route("/", get(root))
        .nest_service("/home", get_service(ServeDir::new("assets")));    // <<------ 
    let api_server_addr = "localhost:3000";

    // Create a TCP listener
    let listener = tokio::net::TcpListener::bind(api_server_addr).await?;

    // Start serving the app
    println!("API server listening on: {api_server_addr}");
    axum::serve(listener, routes_all).await?;

    Ok(()) // main() completed with no errors
}

// endregion: ---- MAIN ----
async fn root() -> Html<String> {
    Html("<h1> Hello from the root route</h1>".to_string())
}
```
=== assets/index.html ===
```html
<!DOCTYPE html>
<html lang="en">

<head>
    <title>Test</title>
    <meta charset='utf-8'>
    <meta name="viewport" content="width=device-width, initial-scale=1">
</head>

<body>
    <p>Hello from index.html</p>
    <script src="my-script.js"></script>   <!-- Cant find this file -->
</body>

</html>
```

### axum version

axum = "0.7.5", tower-http = {version = "0.5.2", features = ["fs"]}

**Comment by jplatte** [comment]
How are you testing this? If you're opening the page in a browser at `http://127.0.0.1:3000/home/index.html`, it should try to load the script from `http://127.0.0.1:3000/home/my-script.js` (and find it), rather than `http://127.0.0.1:3000/my-script.js` as the error message suggests

**Reply by xabra** [reply]
**jplatte**: Thanks for looking at this.  You make an interesting observation.

I test by opening the page **http://127.0.0.1:3000/home**
Axum (or tower-http) automatically finds the _index.html_ file in the physical linux directory _.../.../.../my-project/assets_ 
My understanding is that this code:
```rust
    let routes_all = Router::new()
        .route("/", get(root))
        .nest_service("/home", get_service(ServeDir::new("assets")));      // <<------ 
```
will serve the contents of the _assets_ **directory** on a **route** named _/home_ .  
It seems to do this correctly for _index.html_, but not _my-script.js_ which is in the same directory and referenced from _index.html_

However, your observation about the error message prompted me to try serving the directory from the root route '/', like this:
```rust
    let routes_all = Router::new()
        //.route("/", get(root))   // Commented out to avoid conflicting
        .nest_service("/", get_service(ServeDir::new("assets")));   // Moved from route  "/home" to  "/"
```
And that works.
Is this a bug in Axum or tower-http?

**Reply by mladedav** [reply, marked as answer]
I suspect this could be another case of the `/home/` vs `/home` nesting issue.

When you called `/home`, the directory is `/` so the browser tried to fetch `/my-script.js`. If you had used `nest_service("/home/", ...);` and called `localhost/home/` instead, I think it would correctly use the relative path as `/home/my-script.js`.

You could try calling `locahost/home/index.html`, I think that version should work. I didn't try to run any code though, so I'm just extrapolating and guessing here.

**Reply by xabra** [reply]
Ok - resolved (sort of).  
Here is my understanding now:

Suppose one has a directory containing an _index.html_ file which has links to other files in that directory such as stylesheets, javascript, images, etc. 
And further suppose one is using _nest_service_ (and _ServeDir_) to serve that directory at some route.  
Then, the paths in those links must be modified to include the route.

For example, if  _assets/index.html_ references _assets/my-script.js_ like this: 
```html
<script src="my-script.js"></script> 
```
and we use nest_service to serve the _assets_ directory on the route _/home_ like this:
```rust
let routes_all = Router::new()
        .route("/", get(root))
        .nest_service("/home", get_service(ServeDir::new("assets"))); 
```
then one must modify the tag to this:
```html
<script src="home/my-script.js"></script>
```
My code works now, but to me it doesn't make sense that this should be the way nest_service/ServeDir work.
It is not portable:  if you decide to nest the directory under a different route, then all of the internal html references have to be changed.
I would have thought the nesting would implicitly include the route, as it does for the _index.html_ file

**Reply by xabra** [reply]
Ignore my reply directly above this.
**mladedav** figured it out.  Thank you!

There is no need for any changes to the paths to files in the html.
Using:
```rust 
.nest_service("/home/", get_service(ServeDir::new("assets")));
```
and browsing to _http://127.0.0.1:3000/home/_ solves the problem
